### PR TITLE
Feature/improve technology routes

### LINKED ISSUE #148: Ajustar rota de cadastro de tecnologia
## Feature Description

Ajustar o fluxo de cadastro de tecnologia na API para:
- Associar termos à tecnologia;
- Indexar os termos da taxonomia de categoria, juntamente com os dados da tecnologia, no algolia.

---------------

## Acceptance criteria
- Se um `term`, pertencente a qualquer taxonomia, for enviado na rota de cadastro de tecnologia, ele deve ser associado àquela tecnologia na nossa base de dados;
- Se um `term` pertencente à taxonomia `Category` for enviado na rota de cadastro de tecnologia, ele deve ser indexado no algolia;
- Se um `term` **não** pertencente à taxonomia `Category`, for enviado na rota de cadastro de tecnologia, **não** deve ser indexado no algolia;

## Implementation Brief

- A rota de cadastro (`store`)  da tecnologia, deve aceitar, **de maneira opcional**, o envio de um array de `termId` ou `slug`. Ex.:
```
terms: ['Saneamento', 10]
```
Caso essa propriedade seja enviada, para cada um desses slugs ou termIds deve ser criada uma relação entre tecnologia e term na tabela `technology_term`
- Caso o array seja enviado e um dos `terms` pertença à `taxonomy` cujo slug é `CATEGORY`, ele deve ser associado no algolia como uma propriedade chamada `category`, juntamente com as demais propriedades da tecnologia:
```
...technologyData,
category: term.term
```

### PR DESCRIPTION
**Pull request by lcnogueira**
## Summary

Resolves #148 

## Relevant technical choices
- A atualização dos `terms` de uma tecnologia funciona do mesmo modo que a atualização de `users` da tecnologia. Ou seja, sempre ocorre um detach de todos os terms e depois um attach de todos os terms. Ex.: se uma tecnologia tem os `terms` [1,2] e queremos deletar só 1 e inserir 2, deve ser enviada uma lista de terms [2,3]. 
- A indexação dos dados de tecnologia no algolia deixou de ser realizada dentro do hook `afterSave` do model `technology`. Isso porque, no momento em que o hook é executado, os `terms` associados à tecnologia ainda não estão na base de dados, não permitindo assim indexar a categoria (caso exista) da tecnologia no algolia. Essa indexação é feita (por enquanto) dentro do próprio controller.
- Agora a rota de atualização de tecnologias aceita um array de `terms` (e não mais um único `term`), assim como a rota de cadastro de tecnologia.
- O cadastro e atualização das tecnologias agora é feito em uma transação;

## Outros ajustes
- Foram realizados ajustes nas rotas das tecnologias que retormavam um array com 1 tecnologia para que retornassem um objeto, visto que o retorno esperado era sempre de uma única tecnologia;
- O campo `password` do model de `User` foi transformado em `hidden` para não ser retornado durante os fetchs;

## QA Steps
- Utilize a rota de cadastro de tecnologia (post `/technologies`) e envie um array com `terms` que possua slug e/ou id dos terms. **Insira pelo menos um dos termos da taxonomia de Categoria**. Ex.: `terms: ['Saneamento', 10]`. Em seguida:
  - Verifique no banco ( `technologies` e `technology_terms`) que os dados foram inseridos;
  - Verifique no algolia que a tecnologia foi inserida e existe um campo `category` que tem agora o term da taxonomia de categoria (nesse caso, `Saneamento`).
- Utilize a rota de cadastro de tecnologia e envie um array com `terms` que **não** seja da taxonomia de categoria **OU** sem os terms. Em seguida:
  - Verifique no banco ( `technologies` e `technology_terms`) que os dados foram inseridos; 
  - Verifique no algolia que a tecnologia foi inserida e existe um campo `category` que tem agora uma string padrão: "Não definida".
- Faça exatamene os testes acima porém utilizando a rota de **atualização da tecnologia** (put `/technologies/:id`)

## Checklist

- [X] My code is tested and passes existing unit tests.
- [X] My code has an appropriate set of unit tests which all pass.
- [X] My code passes the linting.
- [X] My code has proper inline documentation.
- [X] I have manually tested this PR.

### CODE REVIEW

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 1]
```diff
@@ -1,11 +1,18 @@
 /** @typedef {import('@adonisjs/framework/src/Request')} Request */
 /** @typedef {import('@adonisjs/framework/src/Response')} Response */
 
+const Config = use('Adonis/Src/Config');
```
Comment (mateus4k) [inline packages/api/app/Controllers/Http/TechnologyController.js:4]: Pode substituir pelo alias, se quiser :smile:

```js
const Config = use('Config')
```

**Comment by lcnogueira** [comment]
@alexandreadames , mexi nas rotas de tecnologia. Tentei descrever o máximo aqui na PR. Seria interessante dar uma revisada, já que vc trabalhou/trabalha nessa parte :+1:

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 2]
```diff
@@ -190,38 +235,36 @@ class TechnologyController {
 		const { idTechnology } = params;
 		const technology = await Technology.findOrFail(idTechnology);
 		await this.syncronizeUsers(users, technology);
-		return Technology.query()
-			.with('users')
-			.where('id', technology.id)
-			.fetch();
+		await technology.load('users');
+		return technology;
```
Comment (nicholasio) [inline packages/api/app/Controllers/Http/TechnologyController.js:239]:
```suggestion
		return technology.load('users');
```

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 3]
```diff
@@ -190,38 +235,36 @@ class TechnologyController {
...
 	}
 
 	/**
 	 * Update technology details.
 	 * PUT or PATCH technologies/:id
-	 * If termId or termSlug is passed in body, creates a new technolgy term
-	 * If users is passed updates users in technology
+	 * If terms is provided, it updates the related terms
```
Comment (nicholasio) [inline packages/api/app/Controllers/Http/TechnologyController.js:245]:
```suggestion
	 * If terms are provided, the related terms are updated.
```

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 4]
```diff
@@ -190,38 +235,36 @@ class TechnologyController {
...
+	 * If users is provided, it updates the related users
```
Comment (nicholasio) [inline packages/api/app/Controllers/Http/TechnologyController.js:246]:
```suggestion
	 * If users are provided, the related users are updated.
```

File: packages/api/app/Models/User.js [hunk 5]
```diff
@@ -28,6 +28,10 @@ class User extends Model {
 		return ['full_name'];
 	}
 
+	static get hidden() {
```
Comment (nicholasio) [inline packages/api/app/Models/User.js:31]: Boa 👍

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 6]
```diff
@@ -166,21 +173,59 @@ class TechnologyController {
...
+		for (const term of terms) {
+			const termPromise = Term.getTerm(term);
+			termPromises.push(termPromise);
+		}
+		const termInstances = await Promise.all(termPromises);
+		await technology.terms().saveMany(termInstances);
+	}
+
+	indexToAlgolia(technologyData) {
+		const defaultCategory = 'Não definida';
```
Comment (nicholasio) [inline packages/api/app/Controllers/Http/TechnologyController.js:190]: Eu não sei se precisamos definir uma categoria padrão... se uma tecnologia por ventura não tiver nenhuma categoria, podemos simplesmente indexar ela sem categoria.

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 7]
```diff
@@ -166,21 +173,59 @@ class TechnologyController {
...
+	async syncronizeTerms(terms, technology, detach = false) {
+		if (detach) {
+			await technology.terms().detach();
+		}
+		const termPromises = [];
+		for (const term of terms) {
+			const termPromise = Term.getTerm(term);
+			termPromises.push(termPromise);
+		}
+		const termInstances = await Promise.all(termPromises);
```
Comment (nicholasio) [inline packages/api/app/Controllers/Http/TechnologyController.js:185]:
```suggestion
		const termInstances = await Promise.all(terms.map((term) => Term.getTerm(term)));
```

File: packages/api/app/Controllers/Http/TechnologyController.js [hunk 8]
```diff
@@ -166,21 +173,59 @@ class TechnologyController {
 		});
 	}
 
+	async syncronizeTerms(terms, technology, detach = false) {
```
Comment (nicholasio) [inline packages/api/app/Controllers/Http/TechnologyController.js:176]: O que acontece se eu tentar criar/atualizar uma tecnologia com um termo que não existe? Pelo que estou entendo ele vai criar/atualizar a tecnologia mas lançar uma excessão se o termo não existir.

Acho que vamos ter que fazer isso em uma transação para facilitar as coisas, pois se algum erro for lançado fazemos um fallback na transação, o que acha?

File: packages/api/app/Controllers/Http/TechnologyController.js (hunk 7 above)
Comment (lcnogueira) [reply packages/api/app/Controllers/Http/TechnologyController.js:185]: Bem mais simples :+1:

File: packages/api/app/Controllers/Http/TechnologyController.js (hunk 8 above)
Comment (lcnogueira) [reply packages/api/app/Controllers/Http/TechnologyController.js:176]: Tem razão! Não tinha atentado para isso. Até testei aqui só para confirmar e, de fato, isso que acontece. Para eu lembrar depois:
- Colocar cadastro e atualização em uma transação;
- Atualizar o teste de cadastro e atualização de tecnologia para garantir que a tecnologia não é criada (atualmente, só verificamos o status 400 e a msg de retorno)

File: packages/api/app/Controllers/Http/TechnologyController.js (hunk 6 above)
Comment (lcnogueira) [reply packages/api/app/Controllers/Http/TechnologyController.js:190]: Eu acho que é interessante termos (mesmo que seja outro nome...) para permitir ao usuário filtrar lá na busca. Daí ele vai poder filtrar todas as tecnologias que não possuam categoria definida. O que acha?
filtro

Comment (nicholasio) [reply packages/api/app/Controllers/Http/TechnologyController.js:190]: Podemos deixar assim!
